platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/GossipEvent.java

```diff
@@ -151,22 +150,32 @@ public void setRoundCreated(final long roundCreated) {
     }
 
     /**
-     * Set the id of the node that this event was received from
+     * Perform the necessary actions for this event to be tracked through the intake pipeline.
      *
-     * @param senderNodeId the id of the node that this event was received from
+     * @param intakeCounter the atomic integer tracking the number of events received from this sender which are
+     *                      currently in the intake pipeline
      */
-    public void setSenderNodeId(@NonNull final NodeId senderNodeId) {
-        this.senderNodeId = Objects.requireNonNull(senderNodeId);
+    public void enterIntakePipeline(@NonNull final AtomicInteger intakeCounter) {
+        this.intakeCounter = Objects.requireNonNull(intakeCounter);
+
+        // increment the counter to indicate that this event is now in the intake pipeline
+        intakeCounter.incrementAndGet();
     }
 
     /**
-     * Get the id of the node that this event was received from
-     *
-     * @return the id of the node that this event was received from
+     * Declare that this event is exiting the intake pipeline
      */
-    @Nullable
-    public NodeId getSenderNodeId() {
-        return senderNodeId;
+    public void exitIntakePipeline() {
+        // if the intake counter hasn't been set, then this event wasn't received through gossip
+        if (intakeCounter == null) {
+            return;
+        }
+
+        if (intakeCounter.getAndUpdate(count -> count > 0 ? count - 1 : 0) == 0) {
+            throw new IllegalStateException(
+                    "Event processed from peer, but no known events from that peer were in the intake pipeline."
+                            + "This shouldn't be possible.");
+        }
```

Comment:
This will create a new lambda object each time it's called, which can be expensive (with current GC at least). 

Define this:

```
private static final IntUnaryOperator EXIT_INTAKE = count -> count > 0 ? count - 1 : 0;
```

Then you can call the following without having to create a new lambda each time.

```
        if (intakeCounter.getAndUpdate(EXIT_INTAKE) == 0) {
```